Create model with the help of Python

In [1]:
#!pip3 install pandas --upgrade 
#!pip3 install folium --upgrade
#!pip3 install matplotlib --upgrade
#!pip3 install sklearn --upgrade
#!pip install lxml
#!pip3 install html5lib
#!pip3 install lambda
#!pip3 install json
#!pip3 install arcgis
import pandas as pd
import requests
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors
import folium
from sklearn.cluster import KMeans

Python Package, libraries Used in this Notebook:

    Pandas     : 
    folium     : 
    matplotlib : 
    requests   : 
    sklearn    : 
    Lambda     :
    Numpy      :
    arcgis     :
    html5lib   :
    lxml       :
    
Approach:

I will explore two cities 1. London UK 2) Paris France individually, plot the map to show ehe neighbourhoods and then build our model by clustering similar neighbourhoods together.Plot the new map with the clustered neighbourhoods and compare and discuss the findings and insights.

Data Collection:

Since, we are comparing London(UK) and Paris (France) we fetched required data for the cities of London and Paris i.e. postal codes, neighbourhoods and Wards (boroughs) specific to each of the cities. We scraped "List of areas of London" from wikipedia page and Paris data from to take the 2nd table using the following code:
Source 1: https://en.wikipedia.org/wiki/List_of_areas_of_London (London UK)
Source 2: https://www.data.gouv.fr/fr/datasets/r/e88c6fda-1d09-42a0-a069-606d3259114e (Paris France)

In [2]:
url_london_uk = "https://en.wikipedia.org/wiki/List_of_areas_of_London"
wiki_london_url = requests.get(url_london_uk)
wiki_london_data = pd.read_html(wiki_london_url.text)
wiki_london_data = wiki_london_data[1]
wiki_london_data

Location                     London borough       Post town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
526         Woolwich                          Greenwich          LONDON   
527   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
528  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
529          Yeading                         Hillingdon           HAYES   
530         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS grid ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
526              SE18       020    TQ435795  
527               KT4       020    TQ225655  
528               W12       020    TQ225815  
529               UB4       020    TQ115825  
530               UB7       020    TQ063804  

[531 rows x 6 columns]

In [3]:
Wiki_london_uk_data = pd.read_html(wiki_london_url.text)
Wiki_london_uk_data

[                                                   0
 0  Map all coordinates in "Category:Areas of Lond...
 1                 Download coordinates as: KML · GPX,
             Location                     London borough       Post town  \
 0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
 1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
 2          Addington                         Croydon[8]         CROYDON   
 3         Addiscombe                         Croydon[8]         CROYDON   
 4        Albany Park                             Bexley  BEXLEY, SIDCUP   
 ..               ...                                ...             ...   
 526         Woolwich                          Greenwich          LONDON   
 527   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
 528  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
 529          Yeading                         Hillingdon           HAYES   
 

In [4]:
Wiki_london_uk_data = Wiki_london_uk_data[1]
Wiki_london_uk_data

Location                     London borough       Post town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
526         Woolwich                          Greenwich          LONDON   
527   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
528  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
529          Yeading                         Hillingdon           HAYES   
530         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS grid ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
526              SE18       020    TQ435795  
527               KT4       020    TQ225655  
528               W12       020    TQ225815  
529               UB4       020    TQ115825  
530               UB7       020    TQ063804  

[531 rows x 6 columns]

Replace spaces with underscores and numbers the borough

In [5]:
Wiki_london_uk_data.rename(columns=lambda x: x.strip().replace(" ", "_"), inplace=True)
Wiki_london_uk_data

Location                     London borough       Post_town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
526         Woolwich                          Greenwich          LONDON   
527   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
528  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
529          Yeading                         Hillingdon           HAYES   
530         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS_grid_ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
526              SE18       020    TQ435795  
527               KT4       020    TQ225655  
528               W12       020    TQ225815  
529               UB4       020    TQ115825  
530               UB7       020    TQ063804  

[531 rows x 6 columns]

Select columns that are required for our clustering..

In [6]:
df1_uk = Wiki_london_uk_data.drop( [ Wiki_london_uk_data.columns[0], Wiki_london_uk_data.columns[4], Wiki_london_uk_data.columns[5] ], axis=1)

In [7]:
df1_uk.head()

London borough       Post_town Postcode district
0              Bexley, Greenwich [7]          LONDON               SE2
1  Ealing, Hammersmith and Fulham[8]          LONDON            W3, W4
2                         Croydon[8]         CROYDON               CR0
3                         Croydon[8]         CROYDON               CR0
4                             Bexley  BEXLEY, SIDCUP         DA5, DA14

In [8]:
df1_uk.columns = ['borough','town','post_code']
df1_uk

borough            town  post_code
0                Bexley, Greenwich [7]          LONDON        SE2
1    Ealing, Hammersmith and Fulham[8]          LONDON     W3, W4
2                           Croydon[8]         CROYDON        CR0
3                           Croydon[8]         CROYDON        CR0
4                               Bexley  BEXLEY, SIDCUP  DA5, DA14
..                                 ...             ...        ...
526                          Greenwich          LONDON       SE18
527       Sutton, Kingston upon Thames  WORCESTER PARK        KT4
528             Hammersmith and Fulham          LONDON        W12
529                         Hillingdon           HAYES        UB4
530                         Hillingdon    WEST DRAYTON        UB7

[531 rows x 3 columns]

Remove brackets and numbers from the dataset

In [9]:
df1_uk['borough'] = df1_uk['borough'].map(lambda x: x.rstrip(']').rstrip('0123456789').rstrip('['))
df1_uk

borough            town  post_code
0                Bexley, Greenwich           LONDON        SE2
1    Ealing, Hammersmith and Fulham          LONDON     W3, W4
2                           Croydon         CROYDON        CR0
3                           Croydon         CROYDON        CR0
4                            Bexley  BEXLEY, SIDCUP  DA5, DA14
..                              ...             ...        ...
526                       Greenwich          LONDON       SE18
527    Sutton, Kingston upon Thames  WORCESTER PARK        KT4
528          Hammersmith and Fulham          LONDON        W12
529                      Hillingdon           HAYES        UB4
530                      Hillingdon    WEST DRAYTON        UB7

[531 rows x 3 columns]

Dataframe Dimension

In [10]:
df1_uk.shape

(531, 3)

Filter the data only for London

In [11]:
df1_uk = df1_uk[df1_uk['town'].str.contains('LONDON')]
df1_uk

borough                    town post_code
0                Bexley, Greenwich                   LONDON       SE2
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4
6                              City                  LONDON       EC3
7                       Westminster                  LONDON       WC2
9                           Bromley                  LONDON      SE20
..                              ...                     ...       ...
521                       Redbridge                  LONDON  IG8, E18
522       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8
525                          Barnet                  LONDON       N12
526                       Greenwich                  LONDON      SE18
528          Hammersmith and Fulham                  LONDON       W12

[308 rows x 3 columns]

In [12]:
df1_uk.shape

(308, 3)

There are only 308 rows now I can proceed with getting descriptive statistics

In [13]:
df1_uk.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 308 entries, 0 to 528
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   borough    308 non-null    object
 1   town       308 non-null    object
 2   post_code  308 non-null    object
dtypes: object(3)
memory usage: 9.6+ KB


Geolocations of the London Neighbourhoods

ArcGis API- to plot the map I need geographical co-ordinates. I will use the arcgis package tgo perform this task

In [14]:
from arcgis.geocoding import geocode
from arcgis.gis import GIS
gis = GIS()

London arcgis geocode function to return latitude and longitude

In [15]:
def get_x_y_uk(address1):
   lat_coords = 0
   lng_coords = 0
   g = geocode(address='{}, London, England, GBR'.format(address1))[0]
   lng_coords = g['location']['x']
   lat_coords = g['location']['y']
   return str(lat_coords) +","+ str(lng_coords)

In [16]:
c = get_x_y_uk('WC2')

In [17]:
c

'51.51651000000004,-0.11967999999995982'

In [18]:
geo_coordinates_uk = df1_uk['post_code']    
geo_coordinates_uk

0           SE2
1        W3, W4
6           EC3
7           WC2
9          SE20
         ...   
521    IG8, E18
522         IG8
525         N12
526        SE18
528         W12
Name: post_code, Length: 308, dtype: object

In [19]:
coordinates_latlng_uk = geo_coordinates_uk.apply(lambda x: get_x_y_uk(x))
coordinates_latlng_uk

0       51.492450000000076,0.12127000000003818
1        51.51324000000005,-0.2674599999999714
6       51.51200000000006,-0.08057999999994081
7       51.51651000000004,-0.11967999999995982
9       51.41009000000008,-0.05682999999993399
                        ...                   
521    51.589770000000044,0.030520000000024083
522      51.50642000000005,-0.1272099999999341
525     51.615920000000074,-0.1767399999999384
526      51.48207000000008,0.07143000000002075
528      51.50645000000003,-0.2369099999999662
Name: post_code, Length: 308, dtype: object

Extracting latitude using coordinates

In [20]:
lat_london_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[0])
lat_london_uk

0      51.492450000000076
1       51.51324000000005
6       51.51200000000006
7       51.51651000000004
9       51.41009000000008
              ...        
521    51.589770000000044
522     51.50642000000005
525    51.615920000000074
526     51.48207000000008
528     51.50645000000003
Name: post_code, Length: 308, dtype: object

Extracting longtitute using coordinates

In [21]:
lng_london_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[1])
lng_london_uk

0       0.12127000000003818
1       -0.2674599999999714
6      -0.08057999999994081
7      -0.11967999999995982
9      -0.05682999999993399
               ...         
521    0.030520000000024083
522     -0.1272099999999341
525     -0.1767399999999384
526     0.07143000000002075
528     -0.2369099999999662
Name: post_code, Length: 308, dtype: object

Coordinates of London Neighbourhoods are ready!

In [22]:
london_uk_merged = pd.concat([df1_uk,lat_london_uk.astype(float), lng_london_uk.astype(float)], axis=1)
london_uk_merged.columns= ['borough','town','post_code','latitude','longitude']
london_uk_merged

borough                    town post_code  \
0                Bexley, Greenwich                   LONDON       SE2   
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4   
6                              City                  LONDON       EC3   
7                       Westminster                  LONDON       WC2   
9                           Bromley                  LONDON      SE20   
..                              ...                     ...       ...   
521                       Redbridge                  LONDON  IG8, E18   
522       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8   
525                          Barnet                  LONDON       N12   
526                       Greenwich                  LONDON      SE18   
528          Hammersmith and Fulham                  LONDON       W12   

     latitude  longitude  
0    51.49245    0.12127  
1    51.51324   -0.26746  
6    51.51200   -0.08058  
7    51.51651   -0.11968  
9    51.41009   -0.05683  
..        ...        ...  
521  51.58977    0.03052  
522  51.50642   -0.12721  
525  51.61592   -0.17674  
526  51.48207    0.07143  
528  51.50645   -0.23691  

[308 rows x 5 columns]

In [23]:
london_uk_merged.dtypes

borough       object
town          object
post_code     object
latitude     float64
longitude    float64
dtype: object

Collect coordinates for London to help visualize it on the map

In [24]:
london = geocode(address='London, England, GBR')[0]
london_long_coords = london['location']['x']
london_lati_coords = london['location']['y']

In [25]:
london_long_coords

-0.1272099999999341

In [26]:
london_lati_coords

51.50642000000005

Use of Folium package to visualize the Map of London and the neighbourhoods.

In [27]:
# Creating the map of London
map_London = folium.Map(location=[london_lati_coords, london_long_coords], zoom_start=12)
map_London

# adding markers to map
for latitude, longitude, borough, town in zip(london_uk_merged['latitude'], london_uk_merged['longitude'], london_uk_merged['borough'], london_uk_merged['town']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_London)  
map_London

Venues in London - define Foursquare API credentials to get the venue, categories and neighbourhood in London.

In [28]:
CLIENT_ID = 'UMTRLJPZECG3LDCGBRLRYQXKNONMMHAVL0YGY13JRQGKFCCE' 
CLIENT_SECRET = 'QKUW4EZER2STWF45CEV0D3I4MHA4FSTTUS1YXPTZWDFYW3MW'
VERSION = '20180605' # Foursquare API version
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: UMTRLJPZECG3LDCGBRLRYQXKNONMMHAVL0YGY13JRQGKFCCE
CLIENT_SECRET:QKUW4EZER2STWF45CEV0D3I4MHA4FSTTUS1YXPTZWDFYW3MW


Function to get the neraby venues in the neighbourhood

In [29]:
LIMIT=100
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius,
            LIMIT
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    return(nearby_venues)

Venues in London

In [30]:
venues_in_London = getNearbyVenues(london_uk_merged['borough'], london_uk_merged['latitude'], london_uk_merged['longitude'])

Bexley, Greenwich 
Ealing, Hammersmith and Fulham
City
Westminster
Bromley
Islington
Islington
Barnet
Enfield
Wandsworth
Southwark
City
Richmond upon Thames
Barnet
Islington
Wandsworth
Westminster
Bromley
Newham
Ealing
Westminster
Lewisham
Camden
Southwark
Tower Hamlets
Bexley
City
Lewisham
Greenwich
Tower Hamlets
Camden
Haringey
Tower Hamlets
Haringey
Barnet
Brent
Lambeth
Lewisham
Tower Hamlets
Kensington and Chelsea, Hammersmith and Fulham
Brent
Barnet
Barnet
Southwark
Tower Hamlets
Camden
Tower Hamlets
Waltham Forest
Newham
Islington
Richmond upon Thames
Lewisham
Camden
Westminster
Greenwich
Kensington and Chelsea
Barnet
Westminster
Lewisham
Waltham Forest
Hounslow, Ealing, Hammersmith and Fulham
Brent
Barnet
Lambeth, Wandsworth
Islington
Barnet
Merton
Barnet
Westminster
Barnet, Brent, Camden
Lewisham
Bexley
Haringey
Bromley
Tower Hamlets
Newham
Hackney
Islington
Southwark
Lewisham
Brent
Southwark
Ealing
Kensington and Chelsea
Wandsworth
Southwark
Barnet
Newham
Richmond upon Thames


In [31]:
venues_in_London.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Bexley, Greenwich                 51.49245                  0.12127   
1  Bexley, Greenwich                 51.49245                  0.12127   
2  Bexley, Greenwich                 51.49245                  0.12127   
3  Bexley, Greenwich                 51.49245                  0.12127   
4  Bexley, Greenwich                 51.49245                  0.12127   

                              Venue     Venue Category  
0                      Lesnes Abbey      Historic Site  
1                       Sainsbury's        Supermarket  
2                              Lidl        Supermarket  
3  Abbey Wood Railway Station (ABW)      Train Station  
4                        Costcutter  Convenience Store

In [32]:
venues_in_London.shape

(10365, 5)

we have scraped 10365 records of venues, venues categories for clustering ..

In [33]:
venues_in_London.groupby('Venue Category').max()

Neighbourhood  Neighbourhood Latitude  \
Venue Category                                                        
Accessories Store               Westminster                51.51656   
Adult Boutique                    Islington                51.52969   
African Restaurant              Westminster                51.52587   
American Restaurant          Waltham Forest                51.61780   
Antique Shop                    Westminster                51.55506   
...                                     ...                     ...   
Wings Joint          Hammersmith and Fulham                51.54187   
Women's Store                   Westminster                51.55457   
Xinjiang Restaurant               Southwark                51.47480   
Yoga Studio                     Westminster                51.55457   
Zoo Exhibit                          Camden                51.53354   

                     Neighbourhood Longitude                        Venue  
Venue Category                                                             
Accessories Store                   -0.11968           James Smith & Sons  
Adult Boutique                      -0.08697  Sh! Women's Erotic Emporium  
African Restaurant                  -0.08808           Red Sea Restaurant  
American Restaurant                  0.02795                  Spielburger  
Antique Shop                        -0.11968     The London Silver Vaults  
...                                      ...                          ...  
Wings Joint                         -0.19795                     Wingmans  
Women's Store                       -0.11478           Vivien of Holloway  
Xinjiang Restaurant                 -0.09313                    Silk Road  
Yoga Studio                         -0.03558                    yogahaven  
Zoo Exhibit                         -0.14606                Penguin Beach  

[299 rows x 4 columns]

One Hot Encoding - calculate top 10 common venues to base similarity and calculate the mean of the grouped venue categories for each of the neighbourhoods.

In [34]:
London_venue_cat = pd.get_dummies(venues_in_London[['Venue Category']], prefix="", prefix_sep="")
London_venue_cat

Accessories Store  Adult Boutique  African Restaurant  \
0                      0               0                   0   
1                      0               0                   0   
2                      0               0                   0   
3                      0               0                   0   
4                      0               0                   0   
...                  ...             ...                 ...   
10360                  0               0                   0   
10361                  0               0                   0   
10362                  0               0                   0   
10363                  0               0                   0   
10364                  0               0                   0   

       American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0                        0             0       0                 0   
1                        0             0       0                 0   
2                        0             0       0                 0   
3                        0             0       0                 0   
4                        0             0       0                 0   
...                    ...           ...     ...               ...   
10360                    0             0       0                 0   
10361                    0             0       0                 0   
10362                    0             0       0                 0   
10363                    0             0       0                 0   
10364                    0             0       0                 0   

       Argentinian Restaurant  Art Gallery  Art Museum  ...  \
0                           0            0           0  ...   
1                           0            0           0  ...   
2                           0            0           0  ...   
3                           0            0           0  ...   
4                           0            0           0  ...   
...                       ...          ...         ...  ...   
10360                       0            0           0  ...   
10361                       0            0           0  ...   
10362                       0            0           0  ...   
10363                       0            0           0  ...   
10364                       0            0           0  ...   

       Vietnamese Restaurant  Warehouse Store  Whisky Bar  Wine Bar  \
0                          0                0           0         0   
1                          0                0           0         0   
2                          0                0           0         0   
3                          0                0           0         0   
4                          0                0           0         0   
...                      ...              ...         ...       ...   
10360                      0                0           0         0   
10361                      0                0           0         0   
10362                      0                0           0         0   
10363                      0                0           0         0   
10364                      0                0           0         0   

       Wine Shop  Wings Joint  Women's Store  Xinjiang Restaurant  \
0              0            0              0                    0   
1              0            0              0                    0   
2              0            0              0                    0   
3              0            0              0                    0   
4              0            0              0                    0   
...          ...          ...            ...                  ...   
10360          0            0              0                    0   
10361          0            0              0                    0   
10362          0            0              0                    0   
10363          0            0              0                    0   
10364          0            

In [35]:
London_venue_cat['Neighbourhood'] = venues_in_London['Neighbourhood'] 
fixed_columns = [London_venue_cat.columns[-1]] + list(London_venue_cat.columns[:-1])
London_venue_cat = London_venue_cat[fixed_columns]

London_venue_cat.head()

Neighbourhood  Accessories Store  Adult Boutique  African Restaurant  \
0  Bexley, Greenwich                   0               0                   0   
1  Bexley, Greenwich                   0               0                   0   
2  Bexley, Greenwich                   0               0                   0   
3  Bexley, Greenwich                   0               0                   0   
4  Bexley, Greenwich                   0               0                   0   

   American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0                    0             0       0                 0   
1                    0             0       0                 0   
2                    0             0       0                 0   
3                    0             0       0                 0   
4                    0             0       0                 0   

   Argentinian Restaurant  Art Gallery  ...  Vietnamese Restaurant  \
0                       0            0  ...                      0   
1                       0            0  ...                      0   
2                       0            0  ...                      0   
3                       0            0  ...                      0   
4                       0            0  ...                      0   

   Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  \
0                0           0         0          0            0   
1                0           0         0          0            0   
2                0           0         0          0            0   
3                0           0         0          0            0   
4                0           0         0          0            0   

   Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0              0                    0            0            0  
1              0                    0            0            0  
2              0                    0            0            0  
3              0                    0            0            0  
4              0                    0            0            0  

[5 rows x 300 columns]

Venue categories mean value - calculate the mean venue categories value in each Neighbourhood

In [36]:
London_grouped = London_venue_cat.groupby('Neighbourhood').mean().reset_index()
London_grouped.head()

Neighbourhood  Accessories Store  Adult Boutique  \
0                 Barnet                0.0             0.0   
1  Barnet, Brent, Camden                0.0             0.0   
2                 Bexley                0.0             0.0   
3      Bexley, Greenwich                0.0             0.0   
4     Bexley, Greenwich                 0.0             0.0   

   African Restaurant  American Restaurant  Antique Shop  Arcade  \
0                 0.0             0.001805           0.0     0.0   
1                 0.0             0.000000           0.0     0.0   
2                 0.0             0.000000           0.0     0.0   
3                 0.0             0.000000           0.0     0.0   
4                 0.0             0.000000           0.0     0.0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  ...  \
0               0.0                 0.00722          0.0  ...   
1               0.0                 0.00000          0.0  ...   
2               0.0                 0.00000          0.0  ...   
3               0.0                 0.00000          0.0  ...   
4               0.0                 0.00000          0.0  ...   

   Vietnamese Restaurant  Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  \
0               0.001805              0.0         0.0       0.0        0.0   
1               0.000000              0.0         0.0       0.0        0.0   
2               0.000000              0.0         0.0       0.0        0.0   
3               0.000000              0.0         0.0       0.0        0.0   
4               0.000000              0.0         0.0       0.0        0.0   

   Wings Joint  Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0          0.0            0.0                  0.0          0.0          0.0  
1          0.0            0.0                  0.0          0.0          0.0  
2          0.0            0.0                  0.0          0.0          0.0  
3          0.0            0.0                  0.0          0.0          0.0  
4          0.0            0.0                  0.0          0.0          0.0  

[5 rows x 300 columns]

top most common venue categories

In [37]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Select top 10 to cluster the neighbourhoods

In [38]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

Top venue categories - top venue categories in London

In [39]:
# create a new dataframe for London
neighborhoods_venues_sorted_london = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_london['Neighbourhood'] = London_grouped['Neighbourhood']

for ind in np.arange(London_grouped.shape[0]):
    neighborhoods_venues_sorted_london.iloc[ind, 1:] = return_most_common_venues(London_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_london.head()

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
0                 Barnet           Coffee Shop                  Café   
1  Barnet, Brent, Camden        Hardware Store           Supermarket   
2                 Bexley           Supermarket         Historic Site   
3      Bexley, Greenwich              Bus Stop     Convenience Store   
4     Bexley, Greenwich            Supermarket         Train Station   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0         Grocery Store              Bus Stop                   Pub   
1        Clothing Store                Bakery     Convenience Store   
2              Platform         Train Station     Convenience Store   
3               Daycare           Golf Course                  Park   
4     Convenience Store              Platform         Historic Site   

  6th Most Common Venue       7th Most Common Venue  \
0    Italian Restaurant                    Pharmacy   
1  Gym / Fitness Center      Okonomiyaki Restaurant   
2           Golf Course  Construction & Landscaping   
3         Historic Site  Construction & Landscaping   
4     Accessories Store                Optical Shop   

           8th Most Common Venue 9th Most Common Venue  10th Most Common Venue  
0                    Supermarket    Turkish Restaurant        Sushi Restaurant  
1                         Office           Opera House            Noodle House  
2                       Bus Stop                  Park  Okonomiyaki Restaurant  
3                    Sports Club                 Plaza   Portuguese Restaurant  
4  Paper / Office Supplies Store  Pakistani Restaurant       Outdoor Sculpture

KMeans (Model Building)- 5 number of clusters.

In [40]:
# set number of clusters
k_num_clusters = 5

London_grouped_clustering = London_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_london_uk = KMeans(n_clusters=k_num_clusters, random_state=0).fit(London_grouped_clustering)
kmeans_london_uk

KMeans(n_clusters=5, random_state=0)

Labele Clustered Data

In [41]:
kmeans_london_uk.labels_

array([0, 4, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0,
       0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0], dtype=int32)

Lable City

In [42]:
neighborhoods_venues_sorted_london.insert(0, 'Cluster Labels', kmeans_london_uk.labels_ +1)

add latitude and longitude for each of the neighborhood for plotting

In [43]:
london_data = london_uk_merged

london_data = london_data.join(neighborhoods_venues_sorted_london.set_index('Neighbourhood'), on='borough')

london_data.head()

borough    town post_code  latitude  longitude  \
0              Bexley, Greenwich   LONDON       SE2  51.49245    0.12127   
1  Ealing, Hammersmith and Fulham  LONDON    W3, W4  51.51324   -0.26746   
6                            City  LONDON       EC3  51.51200   -0.08058   
7                     Westminster  LONDON       WC2  51.51651   -0.11968   
9                         Bromley  LONDON      SE20  51.41009   -0.05683   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               2           Supermarket         Train Station   
1               3         Grocery Store                  Park   
6               1    Italian Restaurant           Coffee Shop   
7               1                 Hotel           Coffee Shop   
9               1           Supermarket     Convenience Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0     Convenience Store              Platform         Historic Site   
1         Train Station     Indian Restaurant        Breakfast Spot   
6                 Hotel  Gym / Fitness Center                   Pub   
7        Sandwich Place                  Café                   Pub   
9                 Hotel         Grocery Store  Fast Food Restaurant   

  6th Most Common Venue          7th Most Common Venue  \
0     Accessories Store                   Optical Shop   
1          Optical Shop  Paper / Office Supplies Store   
6            Restaurant                       Wine Bar   
7    Italian Restaurant                      Hotel Bar   
9                  Park                       Bus Stop   

           8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Paper / Office Supplies Store  Pakistani Restaurant      Outdoor Sculpture  
1           Pakistani Restaurant     Outdoor Sculpture        Other Nightlife  
6                   Cocktail Bar        Sandwich Place            Salad Place  
7                        Theater            Restaurant       Sushi Restaurant  
9           Gym / Fitness Center           Golf Course              Gastropub

Drop all the NaN values

In [44]:
london_data_formated = london_data.dropna(subset=['Cluster Labels'])

Visualize clustered Neighbourhoods - Let's plot the clusters

In [45]:
map_clusters_london = folium.Map(location=[london_lati_coords, london_long_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(london_data_formated['latitude'], london_data_formated['longitude'], london_data_formated['borough'], london_data_formated['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters_london)
        
map_clusters_london

Examining Clusters

Cluster 1

In [46]:
london_data_formated.loc[london_data_formated['Cluster Labels'] == 1, london_data_formated.columns[[1] + list(range(5, london_data_formated.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
6                    LONDON               1    Italian Restaurant   
7                    LONDON               1                 Hotel   
9                    LONDON               1           Supermarket   
10                   LONDON               1           Coffee Shop   
12                   LONDON               1           Coffee Shop   
..                      ...             ...                   ...   
521                  LONDON               1                   Pub   
522  LONDON, WOODFORD GREEN               1                  Café   
525                  LONDON               1           Coffee Shop   
526                  LONDON               1                   Pub   
528                  LONDON               1                   Pub   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6             Coffee Shop                 Hotel  Gym / Fitness Center   
7             Coffee Shop        Sandwich Place                  Café   
9       Convenience Store                 Hotel         Grocery Store   
10                    Pub            Food Truck                  Café   
12                    Pub            Food Truck                  Café   
..                    ...                   ...                   ...   
521                  Café         Grocery Store             BBQ Joint   
522                 Hotel                 Plaza                Garden   
525                  Café         Grocery Store              Bus Stop   
526         Grocery Store              Bus Stop           Coffee Shop   
528           Coffee Shop         Grocery Store                  Café   

    5th Most Common Venue  6th Most Common Venue       7th Most Common Venue  \
6                     Pub             Restaurant                    Wine Bar   
7                     Pub     Italian Restaurant                   Hotel Bar   
9    Fast Food Restaurant                   Park                    Bus Stop   
10                   Park  Vietnamese Restaurant          Italian Restaurant   
12                   Park  Vietnamese Restaurant          Italian Restaurant   
..                    ...                    ...                         ...   
521           Coffee Shop          Metro Station          Seafood Restaurant   
522               Theater                    Pub                      Bakery   
525                   Pub     Italian Restaurant                    Pharmacy   
526     Indian Restaurant     Turkish Restaurant  Construction & Landscaping   
528             Gastropub                   Park                       Hotel   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6            Cocktail Bar        Sandwich Place            Salad Place  
7                 Theater            Restaurant       Sushi Restaurant  
9    Gym / Fitness Center           Golf Course              Gastropub  
10                  Hotel          Cocktail Bar   Gym / Fitness Center  
12                  Hotel          Cocktail Bar   Gym / Fitness Center  
..                    ...                   ...                    ...  
521                  Park              Pharmacy            Pizza Place  
522     Outdoor Sculpture   Monument / Landmark             Restaurant  
525           Supermarket    Turkish Restaurant       Sushi Restaurant  
526           Golf Course                  Park                   Pier  
528       Thai Restaurant           Pizza Place     Italian Restaurant  

[300 rows x 12 columns]

Cluster 2

In [47]:
london_data_formated.loc[london_data_formated['Cluster Labels'] == 2, london_data_formated.columns[[1] + list(range(5, london_data_formated.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
0                 LONDON               2           Supermarket   
45   BEXLEYHEATH, LONDON               2           Supermarket   
124               LONDON               2           Supermarket   
291       LONDON, SIDCUP               2           Supermarket   
505               LONDON               2           Supermarket   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Train Station     Convenience Store              Platform   
45          Historic Site              Platform         Train Station   
124         Historic Site              Platform         Train Station   
291         Historic Site              Platform         Train Station   
505         Historic Site              Platform         Train Station   

    5th Most Common Venue 6th Most Common Venue       7th Most Common Venue  \
0           Historic Site     Accessories Store                Optical Shop   
45      Convenience Store           Golf Course  Construction & Landscaping   
124     Convenience Store           Golf Course  Construction & Landscaping   
291     Convenience Store           Golf Course  Construction & Landscaping   
505     Convenience Store           Golf Course  Construction & Landscaping   

             8th Most Common Venue 9th Most Common Venue  \
0    Paper / Office Supplies Store  Pakistani Restaurant   
45                        Bus Stop                  Park   
124                       Bus Stop                  Park   
291                       Bus Stop                  Park   
505                       Bus Stop                  Park   

     10th Most Common Venue  
0         Outdoor Sculpture  
45   Okonomiyaki Restaurant  
124  Okonomiyaki Restaurant  
291  Okonomiyaki Restaurant  
505  Okonomiyaki Restaurant

Cluster 3

In [48]:
london_data_formated.loc[london_data_formated['Cluster Labels'] == 3, london_data_formated.columns[[1] + list(range(5, london_data_formated.shape[1]))]]

town  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
1  LONDON               3         Grocery Store                  Park   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
1         Train Station     Indian Restaurant        Breakfast Spot   

  6th Most Common Venue          7th Most Common Venue 8th Most Common Venue  \
1          Optical Shop  Paper / Office Supplies Store  Pakistani Restaurant   

  9th Most Common Venue 10th Most Common Venue  
1     Outdoor Sculpture        Other Nightlife

Cluster 4

In [49]:
london_data_formated.loc[london_data_formated['Cluster Labels'] == 4, london_data_formated.columns[[1] + list(range(5, london_data_formated.shape[1]))]]


town  Cluster Labels 1st Most Common Venue  \
377  HARROW, STANMOREEDGWARE, LONDON               4                Bakery   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
377  Health & Beauty Service                   Gym     Accessories Store   

     5th Most Common Venue 6th Most Common Venue  \
377  Performing Arts Venue                  Park   

             7th Most Common Venue 8th Most Common Venue  \
377  Paper / Office Supplies Store  Pakistani Restaurant   

    9th Most Common Venue 10th Most Common Venue  
377     Outdoor Sculpture        Other Nightlife

Cluster 5

In [50]:
london_data_formated.loc[london_data_formated['Cluster Labels'] == 5, london_data_formated.columns[[1] + list(range(5, london_data_formated.shape[1]))]]

town  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
121  LONDON               5        Hardware Store           Supermarket   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
121        Clothing Store                Bakery     Convenience Store   

    6th Most Common Venue   7th Most Common Venue 8th Most Common Venue  \
121  Gym / Fitness Center  Okonomiyaki Restaurant                Office   

    9th Most Common Venue 10th Most Common Venue  
121           Opera House           Noodle House

Exploring Paris
Neighbourhoods of Paris
Data Collection
We read the json data with pandas.

In [51]:
!wget -q -O 'france-data.json' https://www.data.gouv.fr/fr/datasets/r/e88c6fda-1d09-42a0-a069-606d3259114e
print("Data Downloaded!")

Data Downloaded!


In [52]:
paris_france_raw = pd.read_json('france-data.json')
paris_france_raw.head()

datasetid  \
0  correspondances-code-insee-code-postal   
1  correspondances-code-insee-code-postal   
2  correspondances-code-insee-code-postal   
3  correspondances-code-insee-code-postal   
4  correspondances-code-insee-code-postal   

                                   recordid  \
0  2bf36b38314b6c39dfbcd09225f97fa532b1fc45   
1  7ee82e74e059b443df18bb79fc5a19b1f05e5a88   
2  e2cd3186f07286705ed482a10b6aebd9de633c81   
3  868bf03527a1d0a9defe5cf4e6fa0a730d725699   
4  21e809b1d4480333c8b6fe7addd8f3b06f343e2c   

                                              fields  \
0  {'code_comm': '645', 'nom_dept': 'ESSONNE', 's...   
1  {'code_comm': '133', 'nom_dept': 'SEINE-ET-MAR...   
2  {'code_comm': '378', 'nom_dept': 'ESSONNE', 's...   
3  {'code_comm': '243', 'nom_dept': 'SEINE-ET-MAR...   
4  {'code_comm': '003', 'nom_dept': 'VAL-DE-MARNE...   

                                            geometry  \
0  {'type': 'Point', 'coordinates': [2.2517129721...   
1  {'type': 'Point', 'coordinates': [3.0529405055...   
2  {'type': 'Point', 'coordinates': [2.1971816504...   
3  {'type': 'Point', 'coordinates': [2.7097808131...   
4  {'type': 'Point', 'coordinates': [2.3335102498...   

                record_timestamp  
0  2016-09-21T00:29:06.175+02:00  
1  2016-09-21T00:29:06.175+02:00  
2  2016-09-21T00:29:06.175+02:00  
3  2016-09-21T00:29:06.175+02:00  
4  2016-09-21T00:29:06.175+02:00

Paris dataframe that we need

In [53]:
paris_field_data = pd.DataFrame()
for f in paris_france_raw.fields:
    dict_new = f
    paris_field_data = paris_field_data.append(dict_new, ignore_index=True)
paris_field_data.head()

code_arr code_cant code_comm code_dept code_reg  \
0        3        03       645        91       11   
1        3        20       133        77       11   
2        1        09       378        91       11   
3        5        14       243        77       11   
4        3        34       003        94       11   

                              geo_point_2d  \
0  [48.750443119964764, 2.251712972144151]   
1   [48.41256065214989, 3.052940505560729]   
2    [48.52726809075556, 2.19718165044305]   
3  [48.87307018579678, 2.7097808131278462]   
4   [48.80588035965699, 2.333510249842654]   

                                           geo_shape id_geofla insee_com  \
0  {'type': 'Polygon', 'coordinates': [[[2.238024...     16275     91645   
1  {'type': 'Polygon', 'coordinates': [[[3.076046...     31428     77133   
2  {'type': 'Polygon', 'coordinates': [[[2.203466...     30975     91378   
3  {'type': 'Polygon', 'coordinates': [[[2.727542...     17000     77243   
4  {'type': 'Polygon', 'coordinates': [[[2.343851...     32123     94003   

               nom_comm        nom_dept     nom_region  population  \
0  VERRIERES-LE-BUISSON         ESSONNE  ILE-DE-FRANCE        15.5   
1  COURCELLES-EN-BASSEE  SEINE-ET-MARNE  ILE-DE-FRANCE         0.2   
2             MAUCHAMPS         ESSONNE  ILE-DE-FRANCE         0.3   
3       LAGNY-SUR-MARNE  SEINE-ET-MARNE  ILE-DE-FRANCE        20.2   
4               ARCUEIL    VAL-DE-MARNE  ILE-DE-FRANCE        19.5   

  postal_code            statut  superficie  z_moyen  
0       91370    Commune simple       999.0    121.0  
1       77126    Commune simple      1082.0     88.0  
2       91730    Commune simple       313.0    150.0  
3       77400  Chef-lieu canton       579.0     71.0  
4       94110  Chef-lieu canton       232.0     70.0

Feature Selection - take the columns that we require

In [54]:
df_paris_2 = paris_field_data[['postal_code','nom_comm','nom_dept','geo_point_2d']]
df_paris_2

postal_code              nom_comm           nom_dept  \
0          91370  VERRIERES-LE-BUISSON            ESSONNE   
1          77126  COURCELLES-EN-BASSEE     SEINE-ET-MARNE   
2          91730             MAUCHAMPS            ESSONNE   
3          77400       LAGNY-SUR-MARNE     SEINE-ET-MARNE   
4          94110               ARCUEIL       VAL-DE-MARNE   
...          ...                   ...                ...   
1295       77520     CESSOY-EN-MONTOIS     SEINE-ET-MARNE   
1296       93420            VILLEPINTE  SEINE-SAINT-DENIS   
1297       77130         CANNES-ECLUSE     SEINE-ET-MARNE   
1298       78930              VILLETTE           YVELINES   
1299       95270  LE PLESSIS-LUZARCHES         VAL-D'OISE   

                                 geo_point_2d  
0     [48.750443119964764, 2.251712972144151]  
1      [48.41256065214989, 3.052940505560729]  
2       [48.52726809075556, 2.19718165044305]  
3     [48.87307018579678, 2.7097808131278462]  
4      [48.80588035965699, 2.333510249842654]  
...                                       ...  
1295   [48.50730730461658, 3.138844194183689]  
1296   [48.95902025378707, 2.536306342059409]  
1297   [48.36403767307805, 2.990786679832767]  
1298  [48.92627887061508, 1.6937417245662671]  
1299  [49.09572967201378, 2.4547564431234923]  

[1300 rows x 4 columns]

We have managed to collect 1300 records

In [55]:
df_paris = df_paris_2[df_paris_2['nom_dept'].str.contains('PARIS')].reset_index(drop=True)
df_paris

postal_code                  nom_comm nom_dept  \
0        75009   PARIS-9E-ARRONDISSEMENT    PARIS   
1        75002   PARIS-2E-ARRONDISSEMENT    PARIS   
2        75011  PARIS-11E-ARRONDISSEMENT    PARIS   
3        75003   PARIS-3E-ARRONDISSEMENT    PARIS   
4        75006   PARIS-6E-ARRONDISSEMENT    PARIS   
5        75004   PARIS-4E-ARRONDISSEMENT    PARIS   
6        75001  PARIS-1ER-ARRONDISSEMENT    PARIS   
7        75017  PARIS-17E-ARRONDISSEMENT    PARIS   
8        75008   PARIS-8E-ARRONDISSEMENT    PARIS   
9        75013  PARIS-13E-ARRONDISSEMENT    PARIS   
10       75012  PARIS-12E-ARRONDISSEMENT    PARIS   
11       75005   PARIS-5E-ARRONDISSEMENT    PARIS   
12       75019  PARIS-19E-ARRONDISSEMENT    PARIS   
13       75020  PARIS-20E-ARRONDISSEMENT    PARIS   
14       75010  PARIS-10E-ARRONDISSEMENT    PARIS   
15       75016  PARIS-16E-ARRONDISSEMENT    PARIS   
16       75018  PARIS-18E-ARRONDISSEMENT    PARIS   
17       75007   PARIS-7E-ARRONDISSEMENT    PARIS   
18       75015  PARIS-15E-ARRONDISSEMENT    PARIS   
19       75014  PARIS-14E-ARRONDISSEMENT    PARIS   

                               geo_point_2d  
0    [48.87689616237872, 2.337460241388529]  
1    [48.86790337886785, 2.344107166658533]  
2    [48.85941549762748, 2.378741060237548]  
3    [48.86305413181178, 2.359361058970589]  
4    [48.84896809191946, 2.332670898588416]  
5   [48.854228281954754, 2.357361938142205]  
6     [48.8626304851685, 2.336293446550539]  
7    [48.88733716648682, 2.307485559493426]  
8    [48.87252726662346, 2.312582560420059]  
9    [48.82871768452136, 2.362468228516128]  
10   [48.83515623066034, 2.419807034965275]  
11  [48.844508659617546, 2.349859385560182]  
12   [48.88686862295828, 2.384694327870042]  
13   [48.86318677744551, 2.400819826729021]  
14   [48.87602855694339, 2.361112904561707]  
15   [48.86039876035177, 2.262099559395783]  
16  [48.892735074561706, 2.348711933867703]  
17   [48.85608259819694, 2.312438687733857]  
18   [48.84015541860987, 2.293559372435076]  
19   [48.82899321160942, 2.327100883257538]

In [56]:
df_paris.shape

(20, 4)

In [57]:
df_paris.dtypes

postal_code     object
nom_comm        object
nom_dept        object
geo_point_2d    object
dtype: object

Gelocations of the Neighbourhoods of Paris

In [58]:
df_paris['geo_point_2d'][0]

[48.87689616237872, 2.337460241388529]

In [59]:
gp2d_temp = df_paris['geo_point_2d']
gp2d_temp

0      [48.87689616237872, 2.337460241388529]
1      [48.86790337886785, 2.344107166658533]
2      [48.85941549762748, 2.378741060237548]
3      [48.86305413181178, 2.359361058970589]
4      [48.84896809191946, 2.332670898588416]
5     [48.854228281954754, 2.357361938142205]
6       [48.8626304851685, 2.336293446550539]
7      [48.88733716648682, 2.307485559493426]
8      [48.87252726662346, 2.312582560420059]
9      [48.82871768452136, 2.362468228516128]
10     [48.83515623066034, 2.419807034965275]
11    [48.844508659617546, 2.349859385560182]
12     [48.88686862295828, 2.384694327870042]
13     [48.86318677744551, 2.400819826729021]
14     [48.87602855694339, 2.361112904561707]
15     [48.86039876035177, 2.262099559395783]
16    [48.892735074561706, 2.348711933867703]
17     [48.85608259819694, 2.312438687733857]
18     [48.84015541860987, 2.293559372435076]
19     [48.82899321160942, 2.327100883257538]
Name: geo_point_2d, dtype: object

In [60]:
paris_latlng = df_paris['geo_point_2d'].astype('str')

Spliting the geo_point_2d column into latitude and longitude.

In [61]:
paris_lat = paris_latlng.apply(lambda x: x.split(',')[0])
paris_lat = paris_lat.apply(lambda x: x.lstrip('['))
paris_lat

0      48.87689616237872
1      48.86790337886785
2      48.85941549762748
3      48.86305413181178
4      48.84896809191946
5     48.854228281954754
6       48.8626304851685
7      48.88733716648682
8      48.87252726662346
9      48.82871768452136
10     48.83515623066034
11    48.844508659617546
12     48.88686862295828
13     48.86318677744551
14     48.87602855694339
15     48.86039876035177
16    48.892735074561706
17     48.85608259819694
18     48.84015541860987
19     48.82899321160942
Name: geo_point_2d, dtype: object

Longitude

In [62]:
paris_lng = paris_latlng.apply(lambda x: x.split(',')[1])
paris_lng = paris_lng.apply(lambda x: x.rstrip(']'))
paris_lng

0      2.337460241388529
1      2.344107166658533
2      2.378741060237548
3      2.359361058970589
4      2.332670898588416
5      2.357361938142205
6      2.336293446550539
7      2.307485559493426
8      2.312582560420059
9      2.362468228516128
10     2.419807034965275
11     2.349859385560182
12     2.384694327870042
13     2.400819826729021
14     2.361112904561707
15     2.262099559395783
16     2.348711933867703
17     2.312438687733857
18     2.293559372435076
19     2.327100883257538
Name: geo_point_2d, dtype: object

In [63]:
paris_geo_lat  = pd.DataFrame(paris_lat.astype(float))
paris_geo_lat.columns=['Latitude']
paris_geo_lng = pd.DataFrame(paris_lng.astype(float))
paris_geo_lng.columns=['Longitude']

Preparing our combined data by dropping the geo_point_2d column

In [64]:
combined_paris_data = pd.concat([df_paris.drop('geo_point_2d', axis=1), paris_geo_lat, paris_geo_lng], axis=1)
combined_paris_data

postal_code                  nom_comm nom_dept   Latitude  Longitude
0        75009   PARIS-9E-ARRONDISSEMENT    PARIS  48.876896   2.337460
1        75002   PARIS-2E-ARRONDISSEMENT    PARIS  48.867903   2.344107
2        75011  PARIS-11E-ARRONDISSEMENT    PARIS  48.859415   2.378741
3        75003   PARIS-3E-ARRONDISSEMENT    PARIS  48.863054   2.359361
4        75006   PARIS-6E-ARRONDISSEMENT    PARIS  48.848968   2.332671
5        75004   PARIS-4E-ARRONDISSEMENT    PARIS  48.854228   2.357362
6        75001  PARIS-1ER-ARRONDISSEMENT    PARIS  48.862630   2.336293
7        75017  PARIS-17E-ARRONDISSEMENT    PARIS  48.887337   2.307486
8        75008   PARIS-8E-ARRONDISSEMENT    PARIS  48.872527   2.312583
9        75013  PARIS-13E-ARRONDISSEMENT    PARIS  48.828718   2.362468
10       75012  PARIS-12E-ARRONDISSEMENT    PARIS  48.835156   2.419807
11       75005   PARIS-5E-ARRONDISSEMENT    PARIS  48.844509   2.349859
12       75019  PARIS-19E-ARRONDISSEMENT    PARIS  48.886869   2.384694
13       75020  PARIS-20E-ARRONDISSEMENT    PARIS  48.863187   2.400820
14       75010  PARIS-10E-ARRONDISSEMENT    PARIS  48.876029   2.361113
15       75016  PARIS-16E-ARRONDISSEMENT    PARIS  48.860399   2.262100
16       75018  PARIS-18E-ARRONDISSEMENT    PARIS  48.892735   2.348712
17       75007   PARIS-7E-ARRONDISSEMENT    PARIS  48.856083   2.312439
18       75015  PARIS-15E-ARRONDISSEMENT    PARIS  48.840155   2.293559
19       75014  PARIS-14E-ARRONDISSEMENT    PARIS  48.828993   2.327101

Co-ordinates for Paris

In [65]:
paris = geocode(address='Paris, France, FR')[0]
paris_lng_coords = paris['location']['x']
paris_lat_coords = paris['location']['y']
print("The geolocation of Paris: ", paris_lat_coords, paris_lng_coords)

The geolocation of Paris:  48.85341000000005 2.3488000000000397


Visualize the Map of Paris

In [66]:
# Creating the map of Paris
map_Paris= folium.Map(location=[paris_lat_coords, paris_lng_coords], zoom_start=12)
map_Paris

# adding markers to map
for latitude, longitude, borough, town in zip(combined_paris_data['Latitude'], combined_paris_data['Longitude'], combined_paris_data['nom_comm'], combined_paris_data['nom_dept']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='Blue',
        fill=True,
        fill_opacity=0.8
        ).add_to(map_Paris)  
    
map_Paris

Venues in Paris - Get the neaby venues present in each neighbourhood of Paris Using function

In [67]:
venues_in_Paris = getNearbyVenues(combined_paris_data['nom_comm'], combined_paris_data['Latitude'], combined_paris_data['Longitude'])

PARIS-9E-ARRONDISSEMENT
PARIS-2E-ARRONDISSEMENT
PARIS-11E-ARRONDISSEMENT
PARIS-3E-ARRONDISSEMENT
PARIS-6E-ARRONDISSEMENT
PARIS-4E-ARRONDISSEMENT
PARIS-1ER-ARRONDISSEMENT
PARIS-17E-ARRONDISSEMENT
PARIS-8E-ARRONDISSEMENT
PARIS-13E-ARRONDISSEMENT
PARIS-12E-ARRONDISSEMENT
PARIS-5E-ARRONDISSEMENT
PARIS-19E-ARRONDISSEMENT
PARIS-20E-ARRONDISSEMENT
PARIS-10E-ARRONDISSEMENT
PARIS-16E-ARRONDISSEMENT
PARIS-18E-ARRONDISSEMENT
PARIS-7E-ARRONDISSEMENT
PARIS-15E-ARRONDISSEMENT
PARIS-14E-ARRONDISSEMENT


In [68]:
venues_in_Paris.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   
1  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   
2  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   
3  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   
4  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   

                    Venue                 Venue Category  
0                  So Nat  Vegetarian / Vegan Restaurant  
1                     RAP                   Gourmet Shop  
2  Le Bouclier de Bacchus                       Wine Bar  
3              Farine & O                         Bakery  
4     Place Saint-Georges                          Plaza

In [69]:
venues_in_Paris.shape

(1269, 5)

We have 1269 venue records for Paris neighbourhoods- Grouping by Venue Categories

In [70]:
venues_in_Paris.groupby('Venue Category').max()

Neighbourhood  Neighbourhood Latitude  \
Venue Category                                                             
Afghan Restaurant       PARIS-11E-ARRONDISSEMENT               48.859415   
African Restaurant       PARIS-9E-ARRONDISSEMENT               48.876896   
American Restaurant     PARIS-19E-ARRONDISSEMENT               48.892735   
Antique Shop             PARIS-9E-ARRONDISSEMENT               48.876896   
Argentinian Restaurant   PARIS-3E-ARRONDISSEMENT               48.863054   
...                                          ...                     ...   
Wine Bar                 PARIS-9E-ARRONDISSEMENT               48.892735   
Wine Shop                PARIS-3E-ARRONDISSEMENT               48.876029   
Women's Store            PARIS-2E-ARRONDISSEMENT               48.867903   
Zoo                     PARIS-12E-ARRONDISSEMENT               48.835156   
Zoo Exhibit             PARIS-12E-ARRONDISSEMENT               48.835156   

                        Neighbourhood Longitude  \
Venue Category                                    
Afghan Restaurant                      2.378741   
African Restaurant                     2.361113   
American Restaurant                    2.384694   
Antique Shop                           2.337460   
Argentinian Restaurant                 2.359361   
...                                         ...   
Wine Bar                               2.400820   
Wine Shop                              2.361113   
Women's Store                          2.344107   
Zoo                                    2.419807   
Zoo Exhibit                            2.419807   

                                                           Venue  
Venue Category                                                    
Afghan Restaurant                                    Afghanistan  
African Restaurant                             Wally Le Saharien  
American Restaurant                                     Harper's  
Antique Shop                             Hôtel des Ventes Drouot  
Argentinian Restaurant                                     Anahi  
...                                                          ...  
Wine Bar                                               Ô Château  
Wine Shop                                         Trois Fois Vin  
Women's Store                               L'Appartement Sézane  
Zoo                                     Parc zoologique de Paris  
Zoo Exhibit             Grande Serre du Parc Zoologique de Paris  

[201 rows x 4 columns]

One Hot Encoding - We need to Encode our venue categories to get a better result for our clustering

In [71]:
Paris_venue_categories = pd.get_dummies(venues_in_Paris[['Venue Category']], prefix="", prefix_sep="")
Paris_venue_categories

Afghan Restaurant  African Restaurant  American Restaurant  \
0                     0                   0                    0   
1                     0                   0                    0   
2                     0                   0                    0   
3                     0                   0                    0   
4                     0                   0                    0   
...                 ...                 ...                  ...   
1264                  0                   0                    0   
1265                  0                   0                    0   
1266                  0                   0                    0   
1267                  0                   0                    0   
1268                  0                   0                    0   

      Antique Shop  Argentinian Restaurant  Art Gallery  Art Museum  \
0                0                       0            0           0   
1                0                       0            0           0   
2                0                       0            0           0   
3                0                       0            0           0   
4                0                       0            0           0   
...            ...                     ...          ...         ...   
1264             0                       0            0           0   
1265             0                       0            0           0   
1266             0                       0            0           0   
1267             0                       0            0           0   
1268             0                       0            0           0   

      Arts & Crafts Store  Asian Restaurant  Athletics & Sports  ...  \
0                       0                 0                   0  ...   
1                       0                 0                   0  ...   
2                       0                 0                   0  ...   
3                       0                 0                   0  ...   
4                       0                 0                   0  ...   
...                   ...               ...                 ...  ...   
1264                    0                 0                   0  ...   
1265                    0                 0                   0  ...   
1266                    0                 0                   0  ...   
1267                    0                 0                   0  ...   
1268                    0                 0                   0  ...   

      Udon Restaurant  Vegetarian / Vegan Restaurant  Venezuelan Restaurant  \
0                   0                              1                      0   
1                   0                              0                      0   
2                   0                              0                      0   
3                   0                              0                      0   
4                   0                              0                      0   
...               ...                            ...                    ...   
1264                0                              0                      0   
1265                0                              0                      0   
1266                0                              0                      0   
1267                0                              0                      0   
1268                0                              0                      0   

      Video Game Store  Vietnamese Restaurant  Wine Bar  Wine Shop  \
0                    0                      0         0          0   
1                    0                      0         0          0   
2                    0                      0         1          0   
3                    0                      0         0          0   
4                    0                      0         0          0   
...                ...                    ...       ...        ...   
1264                 0          

Adding Neighbourhoods to our Data

In [72]:
Paris_venue_categories['Neighbourhood'] = venues_in_Paris['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [Paris_venue_categories.columns[-1]] + list(Paris_venue_categories.columns[:-1])
Paris_venue_categories = Paris_venue_categories[fixed_columns]

Paris_venue_categories.head()

Neighbourhood  Afghan Restaurant  African Restaurant  \
0  PARIS-9E-ARRONDISSEMENT                  0                   0   
1  PARIS-9E-ARRONDISSEMENT                  0                   0   
2  PARIS-9E-ARRONDISSEMENT                  0                   0   
3  PARIS-9E-ARRONDISSEMENT                  0                   0   
4  PARIS-9E-ARRONDISSEMENT                  0                   0   

   American Restaurant  Antique Shop  Argentinian Restaurant  Art Gallery  \
0                    0             0                       0            0   
1                    0             0                       0            0   
2                    0             0                       0            0   
3                    0             0                       0            0   
4                    0             0                       0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  ...  Udon Restaurant  \
0           0                    0                 0  ...                0   
1           0                    0                 0  ...                0   
2           0                    0                 0  ...                0   
3           0                    0                 0  ...                0   
4           0                    0                 0  ...                0   

   Vegetarian / Vegan Restaurant  Venezuelan Restaurant  Video Game Store  \
0                              1                      0                 0   
1                              0                      0                 0   
2                              0                      0                 0   
3                              0                      0                 0   
4                              0                      0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Women's Store  Zoo  Zoo Exhibit  
0                      0         0          0              0    0            0  
1                      0         0          0              0    0            0  
2                      0         1          0              0    0            0  
3                      0         0          0              0    0            0  
4                      0         0          0              0    0            0  

[5 rows x 202 columns]

Venue categories mean value - Group Neighbourhoods and calculate the mean venue categories value

In [73]:
Paris_grouped = Paris_venue_categories.groupby('Neighbourhood').mean().reset_index()
Paris_grouped.head()

Neighbourhood  Afghan Restaurant  African Restaurant  \
0  PARIS-10E-ARRONDISSEMENT            0.00000                0.02   
1  PARIS-11E-ARRONDISSEMENT            0.02439                0.00   
2  PARIS-12E-ARRONDISSEMENT            0.00000                0.00   
3  PARIS-13E-ARRONDISSEMENT            0.00000                0.00   
4  PARIS-14E-ARRONDISSEMENT            0.00000                0.00   

   American Restaurant  Antique Shop  Argentinian Restaurant  Art Gallery  \
0                  0.0           0.0                     0.0          0.0   
1                  0.0           0.0                     0.0          0.0   
2                  0.0           0.0                     0.0          0.0   
3                  0.0           0.0                     0.0          0.0   
4                  0.0           0.0                     0.0          0.0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  ...  Udon Restaurant  \
0     0.00000                  0.0          0.030000  ...              0.0   
1     0.02439                  0.0          0.024390  ...              0.0   
2     0.00000                  0.0          0.000000  ...              0.0   
3     0.00000                  0.0          0.222222  ...              0.0   
4     0.00000                  0.0          0.000000  ...              0.0   

   Vegetarian / Vegan Restaurant  Venezuelan Restaurant  Video Game Store  \
0                           0.01                    0.0           0.00000   
1                           0.00                    0.0           0.02439   
2                           0.00                    0.0           0.00000   
3                           0.00                    0.0           0.00000   
4                           0.00                    0.0           0.00000   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Women's Store  Zoo  Zoo Exhibit  
0               0.000000   0.02000       0.02            0.0  0.0          0.0  
1               0.048780   0.02439       0.00            0.0  0.0          0.0  
2               0.000000   0.00000       0.00            0.0  0.2          0.2  
3               0.206349   0.00000       0.00            0.0  0.0          0.0  
4               0.000000   0.00000       0.00            0.0  0.0          0.0  

[5 rows x 202 columns]

Top venue categories -  using previously defined function

In [74]:
# create a new dataframe for Paris
neighborhoods_venues_sorted_paris = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_paris['Neighbourhood'] = Paris_grouped['Neighbourhood']

for ind in np.arange(Paris_grouped.shape[0]):
    neighborhoods_venues_sorted_paris.iloc[ind, 1:] = return_most_common_venues(Paris_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_paris.head()

Neighbourhood 1st Most Common Venue  2nd Most Common Venue  \
0  PARIS-10E-ARRONDISSEMENT     French Restaurant                 Bistro   
1  PARIS-11E-ARRONDISSEMENT     French Restaurant             Restaurant   
2  PARIS-12E-ARRONDISSEMENT           Zoo Exhibit                 Bistro   
3  PARIS-13E-ARRONDISSEMENT      Asian Restaurant  Vietnamese Restaurant   
4  PARIS-14E-ARRONDISSEMENT     French Restaurant                 Bistro   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Hotel                  Café           Coffee Shop   
1    Italian Restaurant                  Café           Pastry Shop   
2           Supermarket   Monument / Landmark                   Zoo   
3    Chinese Restaurant       Thai Restaurant     French Restaurant   
4                 Hotel                Bakery             Brasserie   

   6th Most Common Venue 7th Most Common Venue    8th Most Common Venue  \
0            Pizza Place      Asian Restaurant        Indian Restaurant   
1  Vietnamese Restaurant                   Bar                   Bakery   
2      Jewish Restaurant                Museum  New American Restaurant   
3              Juice Bar   Japanese Restaurant                    Hotel   
4      Food & Drink Shop              Tea Room     Fast Food Restaurant   

  9th Most Common Venue  10th Most Common Venue  
0    Italian Restaurant      African Restaurant  
1    Mexican Restaurant   Performing Arts Venue  
2          Noodle House  Okonomiyaki Restaurant  
3              Bus Stop                    Park  
4      Sushi Restaurant             Supermarket

K Means (Model Building) Cluster city of Paris to roughly 5

In [75]:
# set number of clusters
k_num_clusters = 5

Paris_grouped_clustering = Paris_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_Paris = KMeans(n_clusters=k_num_clusters, random_state=0).fit(Paris_grouped_clustering)
kmeans_Paris

KMeans(n_clusters=5, random_state=0)

Labelling Clustered Data

In [76]:
kmeans_Paris.labels_

array([0, 0, 2, 3, 4, 0, 1, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 4, 0],
      dtype=int32)

In [77]:
neighborhoods_venues_sorted_paris.insert(0, 'Cluster Labels', kmeans_Paris.labels_ +1)

Model has labeled the city

Join paris_combined_data with our neighbourhood venues to prepare it for plotting

In [78]:
paris_data = combined_paris_data

paris_data = paris_data.join(neighborhoods_venues_sorted_paris.set_index('Neighbourhood'), on='nom_comm')

paris_data.head()

postal_code                  nom_comm nom_dept   Latitude  Longitude  \
0       75009   PARIS-9E-ARRONDISSEMENT    PARIS  48.876896   2.337460   
1       75002   PARIS-2E-ARRONDISSEMENT    PARIS  48.867903   2.344107   
2       75011  PARIS-11E-ARRONDISSEMENT    PARIS  48.859415   2.378741   
3       75003   PARIS-3E-ARRONDISSEMENT    PARIS  48.863054   2.359361   
4       75006   PARIS-6E-ARRONDISSEMENT    PARIS  48.848968   2.332671   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               1     French Restaurant                 Hotel   
1               1     French Restaurant          Cocktail Bar   
2               1     French Restaurant            Restaurant   
3               1     French Restaurant   Japanese Restaurant   
4               1                Bakery     French Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                Bakery                Bistro   Japanese Restaurant   
1              Wine Bar           Coffee Shop              Creperie   
2    Italian Restaurant                  Café           Pastry Shop   
3                Bakery          Gourmet Shop    Italian Restaurant   
4                 Plaza              Tea Room              Wine Bar   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0               Wine Bar                Lounge          Cocktail Bar   
1     Italian Restaurant                 Hotel                Bakery   
2  Vietnamese Restaurant                   Bar                Bakery   
3            Coffee Shop        Sandwich Place              Wine Bar   
4               Fountain        Chocolate Shop           Pastry Shop   

  9th Most Common Venue  10th Most Common Venue  
0           Cheese Shop    Gym / Fitness Center  
1       Thai Restaurant  Furniture / Home Store  
2    Mexican Restaurant   Performing Arts Venue  
3          Cocktail Bar             Art Gallery  
4            Shoe Store                    Café

Drop all the NaN values to prevent data skew

In [79]:
paris_data_formated = paris_data.dropna(subset=['Cluster Labels'])

Visualizing the clustered neighbourhood - Let's plot the clusters

In [80]:
map_clusters_paris = folium.Map(location=[paris_lat_coords, paris_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(paris_data_formated['Latitude'], paris_data_formated['Longitude'], paris_data_formated['nom_comm'], paris_data_formated['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + ' ' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.8
        ).add_to(map_clusters_paris)
        
map_clusters_paris

Examining our Clusters

Cluster 1

In [81]:
paris_data_formated.loc[paris_data_formated['Cluster Labels'] == 1, paris_data_formated.columns[[1] + list(range(5, paris_data_formated.shape[1]))]]

nom_comm  Cluster Labels 1st Most Common Venue  \
0    PARIS-9E-ARRONDISSEMENT               1     French Restaurant   
1    PARIS-2E-ARRONDISSEMENT               1     French Restaurant   
2   PARIS-11E-ARRONDISSEMENT               1     French Restaurant   
3    PARIS-3E-ARRONDISSEMENT               1     French Restaurant   
4    PARIS-6E-ARRONDISSEMENT               1                Bakery   
5    PARIS-4E-ARRONDISSEMENT               1     French Restaurant   
6   PARIS-1ER-ARRONDISSEMENT               1   Japanese Restaurant   
11   PARIS-5E-ARRONDISSEMENT               1     French Restaurant   
12  PARIS-19E-ARRONDISSEMENT               1     French Restaurant   
13  PARIS-20E-ARRONDISSEMENT               1                Bistro   
14  PARIS-10E-ARRONDISSEMENT               1     French Restaurant   
16  PARIS-18E-ARRONDISSEMENT               1     French Restaurant   
18  PARIS-15E-ARRONDISSEMENT               1    Italian Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Hotel                Bakery                Bistro   
1           Cocktail Bar              Wine Bar           Coffee Shop   
2             Restaurant    Italian Restaurant                  Café   
3    Japanese Restaurant                Bakery          Gourmet Shop   
4      French Restaurant                 Plaza              Tea Room   
5         Ice Cream Shop        Clothing Store                  Park   
6      French Restaurant                 Plaza                 Hotel   
11                 Hotel    Italian Restaurant                Bakery   
12                   Bar                 Hotel                  Café   
13                 Plaza                Bakery     French Restaurant   
14                Bistro                 Hotel                  Café   
16                   Bar            Restaurant                 Plaza   
18                 Hotel     French Restaurant   Lebanese Restaurant   

   5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
0    Japanese Restaurant               Wine Bar                Lounge   
1               Creperie     Italian Restaurant                 Hotel   
2            Pastry Shop  Vietnamese Restaurant                   Bar   
3     Italian Restaurant            Coffee Shop        Sandwich Place   
4               Wine Bar               Fountain        Chocolate Shop   
5                  Hotel       Pedestrian Plaza           Pastry Shop   
6             Art Museum                   Café    Italian Restaurant   
11                 Plaza                    Pub                  Café   
12                Bistro            Supermarket              Beer Bar   
13   Japanese Restaurant                  Hotel                   Bar   
14           Coffee Shop            Pizza Place      Asian Restaurant   
16                Bistro            Pizza Place                 Hotel   
18                Bakery              Brasserie       Thai Restaurant   

   8th Most Common Venue 9th Most Common Venue  10th Most Common Venue  
0           Cocktail Bar           Cheese Shop    Gym / Fitness Center  
1                 Bakery       Thai Restaurant  Furniture / Home Store  
2                 Bakery    Mexican Restaurant   Performing Arts Venue  
3               Wine Bar          Cocktail Bar             Art Gallery  
4            Pastry Shop            Shoe Store                    Café  
5               Wine Bar                Bistro                Creperie  
6                 Bakery                Garden             Coffee Shop  
11           Coffee Shop                   Bar           Historic Site  
12    Seafood Restaurant               Brewery              Steakhouse  
13                  Café    Italian Restaurant       Food & Drink Shop  
14     Indian Restaurant    Italian Restaurant      African Restaurant  
16    Italian Restaurant                  Café       Convenience Store  
18                Bistro            Restaurant            

Cluster 2

In [82]:
paris_data_formated.loc[paris_data_formated['Cluster Labels'] == 2, paris_data_formated.columns[[1] + list(range(5, paris_data_formated.shape[1]))]]

nom_comm  Cluster Labels 1st Most Common Venue  \
15  PARIS-16E-ARRONDISSEMENT               2                  Lake   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
15                 Plaza           Bus Station            Art Museum   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
15                  Pool              Bus Stop     French Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
15         Boat or Ferry                  Park      Afghan Restaurant

Cluster 3

In [83]:
paris_data_formated.loc[paris_data_formated['Cluster Labels'] == 3, paris_data_formated.columns[[1] + list(range(5, paris_data_formated.shape[1]))]]

nom_comm  Cluster Labels 1st Most Common Venue  \
10  PARIS-12E-ARRONDISSEMENT               3           Zoo Exhibit   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
10                Bistro           Supermarket   Monument / Landmark   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
10                   Zoo     Jewish Restaurant                Museum   

      8th Most Common Venue 9th Most Common Venue  10th Most Common Venue  
10  New American Restaurant          Noodle House  Okonomiyaki Restaurant

Cluster 4

In [84]:
paris_data_formated.loc[paris_data_formated['Cluster Labels'] == 4, paris_data_formated.columns[[1] + list(range(5, paris_data_formated.shape[1]))]]

nom_comm  Cluster Labels 1st Most Common Venue  \
9  PARIS-13E-ARRONDISSEMENT               4      Asian Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
9  Vietnamese Restaurant    Chinese Restaurant       Thai Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
9     French Restaurant             Juice Bar   Japanese Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
9                 Hotel              Bus Stop                   Park

Cluster 5

In [85]:
paris_data_formated.loc[paris_data_formated['Cluster Labels'] == 5, paris_data_formated.columns[[1] + list(range(5, paris_data_formated.shape[1]))]]

nom_comm  Cluster Labels 1st Most Common Venue  \
7   PARIS-17E-ARRONDISSEMENT               5     French Restaurant   
8    PARIS-8E-ARRONDISSEMENT               5     French Restaurant   
17   PARIS-7E-ARRONDISSEMENT               5     French Restaurant   
19  PARIS-14E-ARRONDISSEMENT               5     French Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
7                  Hotel    Italian Restaurant                Bakery   
8                  Hotel          Cocktail Bar                   Spa   
17                 Hotel    Italian Restaurant                  Café   
19                Bistro                 Hotel                Bakery   

   5th Most Common Venue 6th Most Common Venue       7th Most Common Venue  \
7                  Plaza            Restaurant                        Café   
8    Corsican Restaurant                   Bar  Modern European Restaurant   
17                 Plaza        History Museum                Cocktail Bar   
19             Brasserie     Food & Drink Shop                    Tea Room   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
7    Japanese Restaurant                Bistro            Pizza Place  
8          Grocery Store             Hotel Bar                Theater  
17            Art Museum          Dessert Shop                 Bistro  
19  Fast Food Restaurant      Sushi Restaurant            Supermarket